cabal-install/src/Distribution/Client/SetupWrapper.hs

```diff
@@ -399,38 +472,65 @@ getSetup verbosity options mpkg = do
 -- | Decide if we're going to be able to do a direct internal call to the
 -- entry point in the Cabal library or if we're going to have to compile
 -- and execute an external Setup.hs script.
-getSetupMethod
+withSetupMethod
   :: Verbosity
   -> SetupScriptOptions
   -> PackageDescription
   -> BuildType
-  -> IO (Version, SetupMethod, SetupScriptOptions)
-getSetupMethod verbosity options pkg buildType'
+  -> AllowInLibrary
+  -> ( forall kind. (Version, SetupMethod kind, SetupScriptOptions ) -> r )
+  -> IO r
+withSetupMethod verbosity options pkg buildType' allowInLibrary with
   | buildType' == Custom
-      || buildType' == Hooks
+      || (buildType' == Hooks && isJust (useLoggingHandle options))
       || maybe False (cabalVersion /=) (useCabalSpecVersion options)
-      || not (cabalVersion `withinRange` useCabalVersion options) =
-      getExternalSetupMethod verbosity options pkg buildType'
-  | isJust (useLoggingHandle options)
-      -- Forcing is done to use an external process e.g. due to parallel
-      -- build concerns.
-      || forceExternalSetupMethod options =
-      return (cabalVersion, SelfExecMethod, options)
-  | otherwise = return (cabalVersion, InternalMethod, options)
-
-runSetupMethod :: WithCallStack (SetupMethod -> SetupRunner)
-runSetupMethod InternalMethod = internalSetupMethod
+      || not (cabalVersion `withinRange` useCabalVersion options)
+      || allowInLibrary == Don'tAllowInLibrary =
+      withExternalSetupMethod
+  | -- TODO: once we refactor the Cabal library to be able to take a logging
+    -- handle as an argument, we will be able to get rid of the self-exec method.
+    -- Tracking ticket: #9987.
+    isJust (useLoggingHandle options) =
+      return $ with (cabalVersion, SelfExecMethod, options)
+  | otherwise
+  = do
+    abiOK <-
+      if buildType' == Hooks
+      then do
+        -- NB: compileExternalSetupMethod compiles the hooks executable.
+        _ <- compileExternalSetupMethod verbosity options pkg Hooks
+        externalHooksABI <- externalSetupHooksABI $ hooksProgFilePath (useWorkingDir options) (useDistPref options)
+        let internalHooksABI = hooksVersion
+        return $ externalHooksABI == internalHooksABI
+      else return True
+    if abiOK
+    then do
+      debug verbosity $ "Using in-library setup method with build-type " ++ show buildType'
+      return $ with (cabalVersion, LibraryMethod, options)
+    else do
+      debug verbosity $ "Hooks ABI mismatch; falling back to external setup method."
+      withExternalSetupMethod
```

Comment:
If I understand correctly, in case of the Hooks build-type, this boils down to:
```
      _ <- compileExternalSetupMethod verbosity options pkg Hooks
      externalHooksABI <- externalSetupHooksABI $ hooksProgFilePath (useWorkingDir options) (useDistPref options)
      let internalHooksABI = hooksVersion
      let abiOk = externalHooksABI == internalHooksABI
      if abiOk
          return $ with (cabalVersion, LibraryMethod, options)
        else do
          withExternalSetupMethod
```

Is there an advantage of using the `LibraryMethod` given we had to compile an external setup anyway just to check the ABI?

Now I am wondering, is this always the case? That is, is there a case where we are sure we can use the library method without compiling the script first?

Reply:
You are right that currently we compile both the external hooks executable and the setup executable here, even though we only need to compile the hooks executable in this code path (falling back to the setup executable when there is an ABI mismatch). I can refactor the code to avoid the compilation of the external setup executable (it makes sense to do so).

I think the existing "cabal version" logic (what's at the top of `withSetupMethod`) effectively guarantees that we will have matching Cabal ABI, but given that we need to compile the external hooks executable anyway it makes sense to do an additional version check as belt-and-braces (mismatched Binary instance errors can be a real nightmare to debug). I don't think there's any other work that we can avoid doing, other than not compiling the Setup executable if we don't need it; we need to compile the hooks executable when using build-type: Hooks.

BTW, there are other benefits to going through the in-library method even if we pessimistically compile the setup executable, for instance we skip re-configuring things that `cabal-install` has already configured.